### PR TITLE
feat(typescript-estree): clarify docs and error for program project without matching TSConfig

### LINKED ISSUE #5753: Docs: Improve error message & docs for "The file does not match your project config"
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

The error message from these lines often shows up when folks open a `*.js` config file not covered by their TSConfig, commonly `.eslintrc.js`.

```
const errorLines = [
  '"parserOptions.project" has been set for @typescript-eslint/parser.',
  `The file does not match your project config: ${path.relative(
    extra.tsconfigRootDir || process.cwd(),
    extra.filePath,
  )}.`,
];
```

We should improve this error message, and add an FAQ entry for it.

### Affected URL(s)

https://typescript-eslint.io/docs/linting/troubleshooting

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #5753
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

This PR does three things:

* It changes the error message from `@typescript-eslint/typescript-estree` -as called by `@typescript-eslint/parser` when `parserOptions.program` is specified but no TSConfig includes the file being linted. Now, the error message is much more clear on what's going on, and what some suggested remediations & further docs reading are
* It clarifies & expands the initial context mentioned in the docs
* It moves those docs from Typed Linting FAQs to the general troubleshooting page
    * I'm not 100% on this, and might be changing my opinion from several months ago... but my current hunch is that it'd be good to have all FAQs in one place

### CODE REVIEW

File: docs/linting/TROUBLESHOOTING.md [hunk 1]
```diff
@@ -28,12 +28,43 @@ If you don't find an existing extension rule, or the extension rule doesn't work
 > We release a new version our tooling every week.
 > _Please_ ensure that you [check our the latest list of "extension" rules](https://typescript-eslint.io/rules/#extension-rules) **_before_** filing an issue.
 
-## I get errors telling me "The file must be included in at least one of the projects provided"
+## I get errors telling me "ESLint was configured to run ... However, that TSConfig does not / none of those TSConfigs include this file"
+
+This error will appear in the beginning of a source file from the combination two things:
```
Comment (a-churchill) [inline docs/linting/TROUBLESHOOTING.md:33]:
```suggestion
This error will appear in the beginning of a source file from the combination of two things:
```

**Review by a-churchill — COMMENTED** [review]
As an end user I think the updated error message + associated docs are very helpful!

File: docs/linting/TROUBLESHOOTING.md [hunk 2]
```diff
@@ -28,12 +28,43 @@ If you don't find an existing extension rule, or the extension rule doesn't work
...
+  - If you **do** want to lint the file with [type-aware linting](./TYPED_LINTING.md):
+    - Check the `include` option of each of the tsconfigs that you provide to `parserOptions.project` - you must ensure that all files match an `include` glob, or else our tooling will not be able to find it.
+    - If your file shouldn't be a part of one of your existing tsconfigs (for example, it is a script/tool local to the repo), then consider creating a new tsconfig (we advise calling it `tsconfig.eslint.json`) in your project root which lists this file in its `include`. For an example of this, you can check out the configuration we use in this repo:
+      - [`tsconfig.eslint.json`](https://github.com/typescript-eslint/typescript-eslint/blob/main/tsconfig.eslint.json)
+      - [`.eslintrc.js`](https://github.com/typescript-eslint/typescript-eslint/blob/main/.eslintrc.js)
+
+See our docs on [type aware linting](./TYPED_LINTING.md#i-get-errors-telling-me-the-file-must-be-included-in-at-least-one-of-the-projects-provided) for more information.
```
Comment (bradzacher) [inline docs/linting/TROUBLESHOOTING.md:62]: With your changes - this link is now a cycle! It links to a section which tells you to visit this section.

Maybe just drop the hash?

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 3]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
-          `You unnecessarily included the extension "${extraExtension}" with the "extraFileExtensions" option. This extension is already handled by the parser by default.`,
-        );
-      }
-    });
+  const describedFilePath = describeFilePath(extra.filePath);
+  const relativeProjects = extra.projects.map(describeFilePath);
+  const describedPrograms =
+    relativeProjects.length === 1
+      ? relativeProjects[0]
+      : `\n${relativeProjects.map(project => `- ${project}`).join('\n')}`;
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:59]: doesn't look like there's a test case for this multiple project config case.
probably worth adding one!

**Review by bradzacher — CHANGES_REQUESTED** [review]
not a full review, just a few comments

File: docs/linting/TROUBLESHOOTING.md [hunk 4]
```diff
@@ -28,12 +28,43 @@ If you don't find an existing extension rule, or the extension rule doesn't work
 > We release a new version our tooling every week.
 > _Please_ ensure that you [check our the latest list of "extension" rules](https://typescript-eslint.io/rules/#extension-rules) **_before_** filing an issue.
 
-## I get errors telling me "The file must be included in at least one of the projects provided"
+## I get errors telling me "ESLint was configured to run ... However, that TSConfig does not / none of those TSConfigs include this file"
+
+This error will appear in the beginning of a source file from the combination of two things:
+
+- The ESLint configuration for the source file specifies at least one TSConfig file in `parserOptions.project`
+- None of those TSConfig files includes the source file being linted
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply docs/linting/TROUBLESHOOTING.md:36]: 👍 i'll update here. It is getting irksome having docs in both the website and repo. I'm ruminating on how we'd want to join the two...

File: packages/typescript-estree/src/create-program/createProjectProgram.ts (hunk 3 above)
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/src/create-program/createProjectProgram.ts:59]: There is: `throws when non of multiple projects include the file`

File: docs/linting/TROUBLESHOOTING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply docs/linting/TROUBLESHOOTING.md:36]: #5824

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 5]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
-          );
-          hasMatchedAnError = true;
-        }
-      } else {
+  const extraFileExtensions = extra.extraFileExtensions || [];
+
+  extraFileExtensions.forEach(extraExtension => {
+    if (!extraExtension.startsWith('.')) {
+      errorLines.push(
+        `Found unexpected extension "${extraExtension}" specified with the "extraFileExtensions" option. Did you mean ".${extraExtension}"?`,
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:70]:
```suggestion
        `Found unexpected extension \`${extraExtension}\` specified with the \`parserOptions.extraFileExtensions\` option. Did you mean \`.${extraExtension}\`?`,
```

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 6]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
+      );
+    }
+    if (DEFAULT_EXTRA_FILE_EXTENSIONS.includes(extraExtension)) {
+      errorLines.push(
+        `You unnecessarily included the extension "${extraExtension}" with the "extraFileExtensions" option. This extension is already handled by the parser by default.`,
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:75]:
```suggestion
        `You unnecessarily included the extension \`${extraExtension}\` with the \`parserOptions.extraFileExtensions\` option. This extension is already handled by the parser by default.`,
```

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 7]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
+      );
+    }
+  });
+
+  const fileExtension = path.extname(extra.filePath);
+  if (!DEFAULT_EXTRA_FILE_EXTENSIONS.includes(fileExtension)) {
+    const nonStandardExt = `The extension for the file (${fileExtension}) is non-standard`;
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:82]:
```suggestion
    const nonStandardExt = `The extension for the file (\`${fileExtension}\`) is non-standard`;
```

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 8]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
+    if (extraFileExtensions.length > 0) {
+      if (!extraFileExtensions.includes(fileExtension)) {
         errorLines.push(
-          `${nonStandardExt}. You should add "parserOptions.extraFileExtensions" to your config.`,
+          `${nonStandardExt}. It should be added to your existing "parserOptions.extraFileExtensions".`,
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:86]:
```suggestion
          `${nonStandardExt}. It should be added to your existing \`parserOptions.extraFileExtensions\`.`,
```

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 9]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
         );
         hasMatchedAnError = true;
       }
-    }
-
-    if (!hasMatchedAnError) {
+    } else {
       errorLines.push(
-        'The file must be included in at least one of the projects provided.',
+        `${nonStandardExt}. You should add "parserOptions.extraFileExtensions" to your config.`,
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:92]:
```suggestion
        `${nonStandardExt}. You should add \`parserOptions.extraFileExtensions\` to your config.`,
```

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 10]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
+    const [describedInclusions, describedSpecifiers] =
+      extra.projects.length === 1
+        ? ['that TSConfig does not', 'that TSConfig']
+        : ['none of those TSConfigs', 'one of those TSConfigs'];
+    errorLines.push(
+      `However, ${describedInclusions} include this file. Either:`,
+      `- Change ESLint's list of included files to not include this file`,
+      `- Change ${describedSpecifiers} to include this file`,
+      `- Create a new TSConfig that includes this file and include it in your parserOptions.project`,
+      `See the TypeScript ESLint docs for more info: https://typescript-eslint.io/docs/linting/troubleshooting#`,
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:108]:
```suggestion
      `See the TypeScript ESLint docs for more info: https://typescript-eslint.io/docs/linting/troubleshooting/`,
```

or did you meant to hash link to a specific FAQ?

File: packages/typescript-estree/src/create-program/createProjectProgram.ts [hunk 11]
```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
...
-    ];
-    let hasMatchedAnError = false;
+  // The file was either matched within the tsconfig, or we allow creating a default program
+  if (astAndProgram || createDefaultProgram) {
+    return astAndProgram;
+  }
 
-    const extraFileExtensions = extra.extraFileExtensions || [];
+  const describeFilePath = (filePath: string): string =>
+    path.relative(extra.tsconfigRootDir || process.cwd(), filePath);
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createProjectProgram.ts:52]: I wonder if we should do like this to make it clearer?

```suggestion
  const describeFilePath = (filePath: string): string => {
    const relative = path.relative(extra.tsconfigRootDir || process.cwd(), filePath);
    if (extra.tsconfigRootDir) {
      return `<tsconfigRootDir>/${relative}`;
    }
    return `<cwd>/${relative}`;
  };
```

**Review by bradzacher — COMMENTED** [review]
summore comments! Looking good!

File: packages/typescript-estree/src/create-program/createProjectProgram.ts (hunk 10 above)
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/src/create-program/createProjectProgram.ts:108]: Oh, right, it'd be good to link to the specific (newly updated) FAQ article.

File: docs/linting/TROUBLESHOOTING.md [hunk 12]
```diff
@@ -28,12 +28,48 @@ If you don't find an existing extension rule, or the extension rule doesn't work
 > We release a new version our tooling every week.
 > _Please_ ensure that you [check our the latest list of "extension" rules](https://typescript-eslint.io/rules/#extension-rules) **_before_** filing an issue.
 
-## I get errors telling me "The file must be included in at least one of the projects provided"
+## I get errors telling me "ESLint was configured to run ... However, that TSConfig does not / none of those TSConfigs include this file"
+
+### Fixing the Error
```
Comment (JoshuaKGoldberg) [inline docs/linting/TROUBLESHOOTING.md:33]: I split this section into these two h3s. Re-reading the un-split explanation felt like a tl;dr.

**Review by bradzacher — APPROVED** [review]
image

**Pull request merged**